experimental/InvariantTheory/primary_invariants.jl

```diff
@@ -0,0 +1,185 @@
+    if !isinteger(c)
+      return false
+    end
+    if c < 0
+      return false
+    end
+  end
+  return true
+end
+
+# Returns possible degrees of primary invariants d_1, ..., d_n with
+# d_1 \cdots d_n == k*|G|, where G = group(R).
+# TODO: Possibly include checks involving earlier runs of the search, see Kemper, p. 181
+function candidates_primary_degrees(R::InvRing, k::Int)
+
+  factors = Vector{fmpz}()
+  for n in [ k, order(group(R)) ]
+    # If we can't factor the group order in reasonable time, we might as well
+    # give up.
+    fac = factor(n) # TODO: cache this!
+    for (p, e) in fac
+      for i = 1:e
+        push!(factors, p)
+      end
+    end
+  end
+
+  # Find all possible tuples (d_1, ..., d_n), such that d_1 \cdots d_n = k*order(group(R))
+  n = degree(group(R))
+  parts = Vector{Any}()
+  for i = 1:n
+    # Actually, we want multiset partitions here (as the factors may show
+    # up more than once). But GAP doesn't have this as far as I'm aware.
+    append!(parts, GAP.gap_to_julia(GAP.Globals.PartitionsSet(GAP.Globals.Set(GAP.julia_to_gap(1:length(factors))), i)))
+  end
+
+  sorted_degrees = Vector{Tuple{fmpz, Vector{fmpz}}}()
+  for part in parts
+    ds = ones(fmpz, n)
+    for i = 1:length(part)
+      for j in part[i]
+       ds[i + n - length(part)] *= factors[j]
+      end
+    end
+    sort!(ds) # To avoid duplicates
+    s = sum(ds)
+    sds = (s, ds)
+    l = searchsortedfirst(sorted_degrees, sds)
+    # If we used multiset partitions there shouldn't be duplicates (I think)
+    if l <= length(sorted_degrees) && sorted_degrees[l] == sds
+      continue
+    end
+    insert!(sorted_degrees, l, sds)
+  end
+
+  degrees = Vector{Vector{fmpz}}()
+  for (s, ds) in sorted_degrees
+    if test_primary_degrees_via_hilbert_series(R, ds)
+      push!(degrees, ds)
+    end
+  end
+  return degrees
+end
+
+# Checks whether there exist homogeneous f_1, ..., f_k in RG of degrees
+# degrees[length(invars) + 1:length(invars) + k] such that
+# RG/< invars, f_1, ..., f_k > has Krull dimension n - k, where n == length(invars).
+# If the base field is finite, the answer "true" might be wrong (for theoretical reasons).
+# See Kem99, Theorem 2.
+function check_primary_degrees(RG::InvRing{FldT, GrpT, PolyElemT}, degrees::Vector{fmpz}, invars::Vector{PolyElemT}, k::Int, iters::Dict{Int, <: VectorSpaceIterator}) where {FldT, GrpT, PolyElemT}
+  R = polynomial_ring(RG)
+  n = length(degrees)
+
+  deg_dict = Dict{fmpz, Int}()
+  for e in degrees[length(invars) + 1:length(invars) + k]
+    deg_dict[e] = get(deg_dict, e, 0) + 1
+  end
+  for degs in Hecke.subsets(Set(keys(deg_dict)))
+    if isempty(invars)
+      I = ideal(R, R())
+    else
+      I = ideal(R, invars)
+    end
+    for e in degs
+      iter = get!(iters, Int(e)) do
+        vector_space_iterator(coefficient_ring(RG), iterate_basis(RG, Int(e)))
+      end
+      I = I + ideal(R, collect_basis(iter))
+    end
+    numbersInds = sum(Int[ deg_dict[e] for e in degs ])
```

Comment:
```suggestion
    numbersInds = sum(deg_dict[e] for e in degs)
```
is slightly faster and uses less memory

Reply:
The problem was that `degs` might be empty and then this would be a sum over an empty collection. But a `?:` solves this problem of course.